File: src/network/layer.cpp
#include "network/layer.hpp"

Layer::Layer(int nin, int nout, std::string activation) : weights_(Matrix(nout, nin, 1)),
                                                          biases_(Matrix(nout, 1, 0.0)),
                                                          activation_(activation)
{
    // Initialise weights.
    float limit = 0;
    if (activation == "sigmoid")
        // Glorot/Xavier initialization.
        limit = sqrt(6.0f / (nin + nout));
    else if (activation == "relu")
        // He initialization.
        limit = sqrt(2.0f / nin);
    else
        // Default initialization.
        limit = 1.0f;
    // Fill weights with random values in range [-limit, limit].
    for (int i = 0; i < nout; i++)
        for (int j = 0; j < nin; j++)
            this->weights_(i, j) = randomRange(-limit, limit);
}

void Layer::pass(Matrix x)
{
    this->x_ = x;
    this->z_ = this->weights_.dot(x).add(this->biases_);

    if (this->activation_ == "sigmoid")
    {
        // Sigmoid function.
        auto sigmoid = [](float x)
        { return 1.0 / (1.0 + std::exp(-x)); };

        // Sigmoid derivative function.
        auto sigmoid_deriv = [](float x)
        {
            float s = 1.0 / (1.0 + std::exp(-x));
            return s * (1.0 - s);
        };

        // Apply both.
        this->a_ = this->z_.apply(sigmoid);
        this->da_ = this->z_.apply(sigmoid_deriv);
    }
    else if (this->activation_ == "relu")
    {
        // ReLU function.
        auto relu = [](float x)
        { return x > 0 ? x : 0; };

        // ReLU derivative function.
        auto relu_deriv = [](float x)
        { return x > 0 ? 1.0 : 0.0; };

        // Apply both.
        this->a_ = this->z_.apply(relu);
        this->da_ = this->z_.apply(relu_deriv);
    }
    else if (this->activation_ == "softmax")
    {

        // Calculate sum of exponentials for softmax.
        float sum = 0.0f;
        for(int i = 0; i < this->z_.rows(); i++)
        {
            sum += std::exp(this->z_(i, 0));
        }

        // Softmax function: a_i = exp(z_i) / sum(exp(z_j))
        // where j iterates over all elements in the row.
        this->a_ = this->z_;
        for(int i = 0; i < this->a_.rows(); i++)
        {
            this->a_(i, 0) = std::exp(this->z_(i, 0)) / sum;
        }

        // Softmax derivative function: da_i = a_i * (1 - a_i)
        this->da_ = this->a_;
        for(int i = 0; i < this->da_.rows(); i++)
        {
            this->da_(i, 0) = this->a_(i, 0) * (1.0 - this->a_(i, 0));
        }
    }

    else // Linear by default.
    {
        this->a_ = this->z_;
        this->da_ = Matrix(this->z_.rows(), this->z_.cols(), 1.0);
    }
}

void Layer::backward(float learning_rate)
{

    Matrix grad_w = this->delta_z.dot(this->x_.transpose());
    Matrix grad_b = this->delta_z;

    this->weights_ = this->weights_.subtract(grad_w.multiply(learning_rate));
    this->biases_ = this->biases_.subtract(grad_b.multiply(learning_rate));
}

Matrix &Layer::weights()
{
    return this->weights_;
}

const Matrix &Layer::weights() const
{
    return this->weights_;
}

Matrix &Layer::biases()
{
    return this->biases_;
}

const Matrix &Layer::biases() const
{
    return this->biases_;
}

void Layer::debugPrint()
{
    std::cout << "l1 weights:" << std::endl;
    this->weights().basicPrint();
    std::cout << "l1 biases:" << std::endl;
    this->biases().basicPrint();
}

File: python/scratchneuralnetwork/layer.py
from .tensor import Tensor
from enum import Enum
import math
import random


class ActiFun(Enum):
    NONE = None
    SIGMOID = 1
    RELU = 2

    def _limit(self, nin: int, nout: int) -> float:
        match self:
            case ActiFun.RELU:
                return math.sqrt(2.0 / nin)
            case _:
                return math.sqrt(6.0 / (nin + nout))

    def forward(self, x: Tensor) -> Tensor:
        match self:
            case ActiFun.SIGMOID:
                return Tensor._from_impl(x._impl.sigmoid_tensor(), None, [x])
            case ActiFun.RELU:
                return Tensor._from_impl(x._impl.relu_tensor(), None, [x])
            case _:
                return x

    def derivative(self, x: Tensor) -> Tensor:
        match self:
            case ActiFun.SIGMOID:
                return Tensor._from_impl(x._impl.sigmoid_derivative_tensor(), None, [x])
            case ActiFun.RELU:
                return Tensor._from_impl(x._impl.relu_derivative_tensor(), None, [x])
            case _:
                return Tensor.ones(*x.shape)


class Layer:
    """Layer class."""

    def __init__(
        self,
        nin: int = 0,
        nout: int = 0,
        acti_fun: ActiFun = ActiFun.NONE,
    ) -> None:

        self.nin: int = nin
        self.nout: int = nout
        self.acti_fun: ActiFun = acti_fun

        limit = acti_fun._limit(nin, nout)
        data = [[random.uniform(-limit, limit) for _ in range(nin)] for _ in range(nout)]
        self.weights: Tensor = Tensor(data)
        self.biases: Tensor = Tensor.zeros(nout, 1)
        self.x: Tensor = None
        self.z: Tensor = None
        self.a: Tensor = None
        self.da: Tensor = None
        self.delta_z: Tensor = None

    def forward(self, tin: Tensor) -> None:
        self.x = tin
        self.z = Tensor._from_impl(
            self.weights._impl.dot_add_bias_tensor(self.x._impl, self.biases._impl),
            None, [self.weights, self.x, self.biases],
        )
        self.a = self.acti_fun.forward(self.z)
        self.da = self.acti_fun.derivative(self.z)

    def backward(self, learning_rate: float) -> None:
        grad_w = self.delta_z @ self.x.transpose()
        grad_b = self.delta_z.sum_cols()

        self.weights = self.weights - grad_w * learning_rate
        self.biases = self.biases - grad_b * learning_rate

    def __repr__(self):
        return f"Layer(nin={self.nin}, nout={self.nout}, acti_fun={self.acti_fun})"
